pyccl/tracers.py

```diff
@@ -0,0 +1,405 @@
+            if kernel_d is None:
+                kernel_d = get_density_kernel(cosmo, dndz)
+            # Transfer (growth rate)
+            z_b, _ = _check_array_params(dndz)
+            a_s = 1./(1+z_b[::-1])
+            t_a = (a_s, -growth_rate(cosmo, a_s))
+            self.add_tracer(cosmo, kernel=kernel_d,
+                            transfer_a=t_a, der_bessel=2)
+        if mag_bias is not None:  # Has magnification bias
+            # Kernel
+            chi, w = get_lensing_kernel(cosmo, dndz, mag_bias=mag_bias)
+            # Multiply by -2 for magnification
+            kernel_m = (chi, -2 * w)
+            self.add_tracer(cosmo, kernel=kernel_m,
+                            der_bessel=-1, der_angles=1)
+
+
+class WeakLensingTracer(Tracer):
+    """Specific `Tracer` associated to galaxy shape distortions including
+    lensing shear and intrinsic alignments within the L-NLA model.
+
+    Args:
+        cosmo (:obj:`Cosmology`): Cosmology object.
+        dndz (tuple of arrays): A tuple of arrays (z, N(z))
+            giving the redshift distribution of the objects. The units are
+            arbitrary; N(z) will be normalized to unity.
+        has_shear (bool): set to `False` if you want to omit the lensing shear
+            contribution from this tracer.
+        ia_bias (tuple of arrays, optional): A tuple of arrays
+            (z, b_IA(z)) giving the intrinsic alignment amplitude b_IA(z).
+            If `None`, the tracer is assumped to not have intrinsic
+            alignments. Defaults to None.
+    """
+    def __init__(self, cosmo, dndz, has_shear=True, ia_bias=None):
+        self.trc = []
+        if has_shear:
+            # Kernel
+            kernel_l = get_lensing_kernel(cosmo, dndz)
+            self.add_tracer(cosmo, kernel=kernel_l,
+                            der_bessel=-1, der_angles=2)
+        if ia_bias is not None:  # Has magnification bias
+            # Kernel
+            kernel_i = get_density_kernel(cosmo, dndz)
+            # Transfer
+            z_a, a = _check_array_params(ia_bias)
+            # Reverse order for increasing a
+            t_a = (1./(1+z_a[::-1]), a[::-1])
+            self.add_tracer(cosmo, kernel=kernel_i, transfer_a=t_a,
+                            der_bessel=-1, der_angles=2)
+
+
+class CMBLensingTracer(Tracer):
+    """A Tracer for CMB lensing.
+    Args:
+        cosmo (:obj:`Cosmology`): Cosmology object.
+        z_source (float): Redshift of source plane for CMB lensing.
+        nsamples (int): number of samples over which the kernel
+            is desired. These will be equi-spaced in radial distance.
+            The kernel is quite smooth, so usually O(100) samples
+            is enough.
+    """
+    def __init__(self, cosmo, z_source, n_samples=100):
+        self.trc = []
+        kernel = get_kappa_kernel(cosmo, z_source, n_samples)
+        self.add_tracer(cosmo, kernel=kernel, der_bessel=-1, der_angles=1)
+
+
+def check_returned_tracer(return_val):
```

Comment:
This function should probably have an `_` in the front of the name so it is private.

Reply:
Done